src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -46,14 +46,25 @@ static auto with(
   kafka::transactional_id tx_id,
   const std::string_view name,
   Func&& func) noexcept {
-    return stm->get_tx_lock(tx_id)->with(
-      [name, tx_id, func = std::forward<Func>(func)]() mutable {
-          vlog(txlog.trace, "got_lock name:{}, tx_id:{}", name, tx_id);
-          return ss::futurize_invoke(std::forward<Func>(func))
-            .finally([name, tx_id]() {
-                vlog(
-                  txlog.trace, "released_lock name:{}, tx_id:{}", name, tx_id);
-            });
+    ss::lw_shared_ptr<mutex> lock = stm->get_tx_lock(tx_id);
+
+    return ss::do_with(
+      std::move(lock),
+      [name, tx_id, func = std::forward<Func>(func)](
```

Comment:
why would we need `do_with` to keep the lock alive if it is already a shared pointer?

Reply:
the idiomatic way to keep a shared_ptr alive when returning is

```
return p->action().finally([p] {});
```